src/natcap/invest/forest_carbon_edge_effect.py

```diff
@@ -804,59 +808,60 @@ def _calculate_tropical_forest_edge_carbon_map(
             row_coords[valid_edge_distance_mask].ravel(),
             col_coords[valid_edge_distance_mask].ravel()))
         # note, the 'n_jobs' parameter was introduced in SciPy 0.16.0
-        distances, indexes = kd_tree.query(
+        # for each forest point x, for each of its k nearest neighbors
+        distances, indexes = kd_tree.query( # distances: (x, k) indexes: (x, k)
             coord_points, k=n_nearest_model_points,
             distance_upper_bound=DISTANCE_UPPER_BOUND, n_jobs=-1)
 
         if n_nearest_model_points == 1:
             distances = distances.reshape(distances.shape[0], 1)
             indexes = indexes.reshape(indexes.shape[0], 1)
 
-        # the 3 is for the 3 thetas in the carbon model
+        # 3 is for the 3 thetas in the carbon model. thetas shape: (x, k, 3)
         thetas = numpy.zeros((indexes.shape[0], indexes.shape[1], 3))
         valid_index_mask = (indexes != kd_tree.n)
         thetas[valid_index_mask] = theta_model_parameters[
             indexes[valid_index_mask]]
 
-        # the 3 is for the 3 models (asym, exp, linear)
-        biomass_model = numpy.zeros(
-            (indexes.shape[0], indexes.shape[1], 3))
         # reshape to an N,nearest_points so we can multiply by thetas
         valid_edge_distances_km = numpy.repeat(
             edge_distance_block[valid_edge_distance_mask] * cell_size_km,
             n_nearest_model_points).reshape(-1, n_nearest_model_points)
 
-        # asymptotic model
+        # For each forest pixel x, for each of its k nearest neighbors, the 
+        # chosen regression method (1, 2, or 3). model_index shape: (x, k)
+        model_index = numpy.empty(indexes.shape, dtype=numpy.int8) 
+        model_index[:] = 0
```

Comment:
If you're making an empty array and then setting them all to zeros, you can do this in one step with a `model_index = numpy.zeros(indexes.shape, dtype=numpy.int8)`

Reply:
I've been using `numpy.empty` because I thought it was faster. I just tested it:
```
>>> timeit.timeit('a = numpy.zeros(100000)', setup='import numpy')
18.696336253000027
>>> timeit.timeit('a = numpy.full(100000, 0)', setup='import numpy')
15.37874967600004
>>> timeit.timeit('a = numpy.empty(100000); a[:] = 0', setup='import numpy')
14.114311920000034
```
And I think we use `numpy.empty` rather than `numpy.full` in a lot of places. However, [this person](https://stackoverflow.com/a/26213344/14451410) got the opposite results. And it doesn't really make sense that `numpy.zeros` would be less optimized. Maybe you have some insight into this?

Reply:
Ha this is straight out of code review I did with James years ago where he was surprised that the empty.. [:] paradigm was faster when i was doing a empty... [:] = nodata set. That's a great answer you linked and and I agree with that person that I'd "use it for the name" and "it's harder for numpy folks to optimize two statements instead of one". Also there's a comment on there about someone guessing that numpy is doing an uninitialized `mmap` for very large arrays, so timing it like this isn't accurate since the memory isn't really set yet. You can see the difference there if you do something that requires reading all the elements of the arrays. Something like this:

```
>>> timeit.timeit('a = numpy.zeros(100000); x=numpy.sum(a)', setup='import numpy')
72.45531159999996
>>> timeit.timeit('a = numpy.empty(100000); a[:] = 0.0; x=numpy.sum(a)', setup='import numpy')
107.3589187
```
(my computer is much slower than yours it seems) 

But even still, in the last statement you're talking about a difference of 0.072 *micro*seconds vs 0.107 microseconds per raster block. Not noticeable on a 4k X 4k raster and taking about 3ish seconds longer on a full global 128k x 64k raster. So even if the results were refersed and empty->[:] were were 3 seconds slower, I think I'd prefer the single statement instead of the two.

But given this I hadn't realized there was a `numpy.full` function! I'll start using that in the future instead of the two step `empty` `[:]=init_val` paradigm that I might have poisoned us with casually back in the day.

Thanks for making the effort on this on what I thought was just a casual style comment.

Reply:
That's wacky! Thanks for the explanation. Weirdly I still get the opposite result:
```
>>> timeit.timeit('a = numpy.zeros(100000); x=numpy.sum(a)', setup='import numpy')
46.84990651900034
>>> timeit.timeit('a = numpy.empty(100000); a[:] = 0.0; x=numpy.sum(a)', setup='import numpy')
43.15799723200007
```
Maybe an operating system difference? But I agree it's worth using `zeros` just for the convenience and that the time difference is insignificant! Fixed in 4175906

Reply:
It might be, at a deep level it's up to the OS about how, or when, or if ever, to allocate memory. I'll merge this PR but I'm curious if you'd be willing to try another experiment? I'll slack you.